lib/kino/smart_cell/chart_builder.ex

```diff
@@ -0,0 +1,206 @@
+    broadcast_event(ctx, "set_axis_options", %{"options" => options})
+  end
+
+  defp convert_field(field, ""), do: {String.to_atom(field), nil}
+
+  defp convert_field(field, value) when field in @as_int do
+    {String.to_atom(field), String.to_integer(value)}
+  end
+
+  defp convert_field(field, value) when field in @as_atom do
+    {String.to_atom(field), String.to_atom(value)}
+  end
+
+  defp convert_field(field, value), do: {String.to_atom(field), value}
+
+  defp vl_alias(%Macro.Env{aliases: aliases}) do
+    Enum.find_value(aliases, fn {current_alias, module} ->
+      if module == VegaLite, do: current_alias
+    end)
+  end
+
+  @impl true
+  def to_attrs(ctx) do
+    ctx.assigns.fields
+    |> Map.put("vl_alias", ctx.assigns.vl_alias)
+  end
+
+  @impl true
+  def to_source(attrs) do
+    attrs
+    |> to_quoted()
+    |> Kino.Utils.Code.quoted_to_string()
+  end
+
+  defp to_quoted(%{"data" => ""}), do: nil
+
+  defp to_quoted(attrs) do
+    data = if attrs["data"], do: String.to_atom(attrs["data"]), else: nil
+    module = if attrs["vl_alias"], do: attrs["vl_alias"], else: VegaLite
+
+    attrs = Map.new(attrs, fn {k, v} -> convert_field(k, v) end)
+
+    [root | nodes] = [
+      %{
+        field: nil,
+        name: :new,
+        module: module,
+        args: [[width: attrs.width, height: attrs.height]]
+      },
+      %{field: :data, name: :data_from_series, module: module, args: [Macro.var(data, nil)]},
+      %{field: :mark, name: :mark, module: module, args: [attrs.chart]},
+      %{
+        field: :x,
+        name: :encode_field,
+        module: module,
+        args: [attrs.x_axis, [type: attrs.x_axis_type]]
+      },
+      %{
+        field: :y,
+        name: :encode_field,
+        module: module,
+        args: [attrs.y_axis, [type: attrs.y_axis_type]]
+      },
+      %{
+        field: :color,
+        name: :encode_field,
+        module: module,
+        args: [attrs.color, [type: attrs.color_type]]
+      }
+    ]
+
+    root = build_root(root)
+
+    nodes
+    |> Enum.map(&clean_node/1)
+    |> Enum.reduce(root, &apply_node/2)
+  end
+
+  defp build_root(root) do
+    args =
+      root.args
+      |> Enum.at(0)
+      |> Enum.filter(&elem(&1, 1))
+      |> case do
+        [] -> []
+        opts -> [opts]
+      end
+
+    quote do
+      unquote(root.module).unquote(root.name)(unquote_splicing(args))
+    end
+  end
+
+  def clean_node(node) do
+    Map.replace!(node, :args, clean_args(node.args))
+  end
+
+  defp apply_node(%{args: nil}, acc), do: acc
+  defp apply_node(%{args: []}, acc), do: acc
+
+  defp apply_node(%{field: field, name: function, module: module, args: args}, acc) do
+    args = if function == :encode_field, do: List.insert_at(args, 0, field), else: args
+
+    quote do
+      unquote(acc) |> unquote(module).unquote(function)(unquote_splicing(args))
+    end
+  end
+
+  defp clean_args(args) do
+    opts =
+      args
+      |> Enum.filter(&is_list/1)
```

Comment:
I think you can simplify if you make `:opts` a separate key in the node, since we know what any options should go as the last argument :)

Reply:
I would propose another approach instead. Instead of having the node do this:

        args: [attrs.color, [type: attrs.color_type]]

I think you should have something like this:

        args: maybe_arg_type(attrs.color, attrs.color_type)

Which builds the correct argument in the first place (or set it to `nil` so it is discarded later).